### ISSUE TITLE
Are fine grained permissions implemented in SQL or Haskell?

### ISSUE THREAD

**Issue by bionicles** [hasura/graphql-engine#3797]
I like how easy it is to define the tables and permissions... just evaluating right now and not using this in production 

To make a medical startup, as a founder, I want to know the data is safe. 

Does Hasura implement the permissions in Postgres with SQL, or in the Hasura server with Haskell?

Just had a nightmare scenario in my mind where, direct connections to the database would not have defined permissions

To put it another way: Is hasura intended to be the only way you connect to a postgres database, and the developer is expected to only grant access to postgres by passing graphql through hasura?

**Comment by coco98** [comment]
@bionicles The permissions are implemented at the Hasura level.

If you're accessing data in Postgres and want to reuse the permissions you've defined, you should go through Hasura and GraphQL. If you're accessing Postgres directly, then the permissions will not be applied.

I wouldn't say that Hasura is intended to be the _only_ way to connect to Postgres. DB admin level tasks, specific use-cases like bulk data migrations etc will be served better by direct postgres access. However, user and app level access should go through Hasura/GraphQL to make sure that the permissions system is used uniformly!

**Comment by bionicles** [comment]
@coco98 @marionschleifer thanks for clarification
